File: 2/2000s/22s/2206_Divide_Array_Into_Equal_Pairs.cpp
/*
https://leetcode.com/problems/divide-array-into-equal-pairs/description/?envType=daily-question&envId=2025-03-17
*/

class Solution {
public:
    bool divideArray(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int sz=nums.size(), nmPairs=sz/2, subPrSz=sz/nmPairs, idx=0;
        pair<int, int>prev;
        while(true){
            if(idx>=sz) break;
            if(prev.first==subPrSz){
                prev.first=0, prev.second=0;
                prev.first++;
                prev.second=nums[idx];
                idx++;
            } else{
                if(prev.first>=1){
                    int tocheck=nums[idx];
                    if(tocheck!=prev.second) return false;
                    prev.first++;
                    idx++;
                } else{
                    prev.first++;
                    prev.second=nums[idx];
                    idx++;
                }
            }
        }
        return true;
    }
};

File: 2/2000s/22s/2206/2206_Divide_Array_Into_Equal_Pairs.java
/*
https://leetcode.com/problems/divide-array-into-equal-pairs/description/?envType=daily-question&envId=2025-03-17
*/

class Pair{

    public int currentPairLength;
    public int lastElement;

    public Pair(int currentPairLength, int lastElement){
        this.currentPairLength=currentPairLength;
        this.lastElement=lastElement;
    }
}


class Solution {
    public boolean divideArray(int[] nums) {
        Arrays.sort(nums);
        int lengthOfNums=nums.length, 
            totalNumPairsFormed=lengthOfNums/2,
            lengthOfEachSubPair=lengthOfNums/totalNumPairsFormed,
            beginIndex=0;
        Pair prev=null;
        while(true){
            if(beginIndex>=lengthOfNums) break;
            if(prev!=null&&prev.currentPairLength==lengthOfEachSubPair){
                prev.currentPairLength=0;
                prev.lastElement=0;
                prev.currentPairLength++;
                prev.lastElement=nums[beginIndex++];
            } else{
                if(prev!=null&&prev.currentPairLength>=1){
                    int toCheckElement=nums[beginIndex++];
                    if(toCheckElement!=prev.lastElement) return false;
                    prev.currentPairLength++;
                } else{
                    prev=new Pair(0, 0);
                    prev.currentPairLength++;
                    prev.lastElement=nums[beginIndex++];
                }
            }
        }
        return true;
    }
}
